### ISSUE TITLE
stale data returned in block-wise CoAP response when multiple GETs performed in one session

### ISSUE THREAD

**Issue by dstabosz** [obgm/libcoap#793]
## Environment

- libcoap version    v4.3.0-67-g997c31f
- Build System:             Cmake
- Operating System:      Linux 
- Operating System Version:   **kernel** 5.4.72-microsoft-standard-WSL2 **distribution** ubuntu 20.04
- Hosted Environment:   None

## Problem Description

When performing a GET to a resource that requires a block-wise response, the first GET works correctly.  However, subsequent GETs to the same URI from the same client session (same source addr/port), libcoap returns the data from the first GET without ever calling the GET handler registered by the application.

Problem seems to be related to lg_xmit not being freed until session times in the server out after 5 minutes.  If another GET comes for the same URI and same client address+port, coap_handle_request_send_block( ) will find the lg_xmit for the first GET and will send it to the client.

I read through the description of #782 and there are some similarities, but they are not the same issue.  I also ran with the latest commit on develop, which has the fix for #782, and I can still recreate the problem.

### Expected Behavior

This example can be run using coap-client and coap-server from libcoap/examples.  The -a and -p options are used below to keep the same session in the server across multiple runs of coap-client.

1. Start the example coap server
- ```coap-server```

2. In a different window, get example_data resource from server
- ```coap-client -a 127.0.0.1 -p 55555 -o example_data.txt -m get coap://127.0.0.1/example_data```
-  data returned from server (note the total size of the resource is 1500 bytes):
```
a123456789b123456789c123456789d123456789e123456789f123456789
...
```
3. Modify the data received from the server (reverse the order of the counting pattern)
- ```tr 123456789 987654321 < example_data.txt > example_data_2.txt```
- Modified data:
```
a987654321b987654321c987654321d987654321e987654321f987654321
...
```
4. Put the modified data to the server:
- ```coap-client -a 127.0.0.1 -p 55555 -f example_data_2.txt -m put coap://127.0.0.1/example_data```
5. Get example_data resource from server.  Expect to get back the modified data:
- ```coap-client -a 127.0.0.1 -p 55555 -m get coap://127.0.0.1/example_data```
- Expected data from server:
```
a987654321b987654321c987654321d987654321e987654321f987654321
...
```

### Actual Behavior

Steps 1 - 4 above work as expected.  However, in step 5 the server returns the original data, not the modified data.

5. Get example_data resource from server.  Expect to get back the modified data:
- ```coap-client -a 127.0.0.1 -p 55555 -m get coap://127.0.0.1/example_data```
- Actual data from server is the original data
```
a123456789b123456789c123456789d123456789e123456789f123456789
...
```

Note that if the client port is changed, the modified data will be returned. 

6. Get example_data resource from server using different client port.  Now modified data will be correctly returned.
- ```coap-client -a 127.0.0.1 -p 55556 -m get coap://127.0.0.1/example_data```
- Actual data from server:
```
a987654321b987654321c987654321d987654321e987654321f987654321
...
```

### Steps to reproduce

```
coap-server &
# Get data from server
coap-client -a 127.0.0.1 -p 55555 -o example_data.txt -m get coap://127.0.0.1/example_data

# modify the data
tr 123456789 987654321 < example_data.txt > example_data_2.txt

# write modified data to server
coap-client -a 127.0.0.1 -p 55555 -f example_data_2.txt -m put coap://127.0.0.1/example_data

# get data from server.  server incorrectly returns the original data instead of the modified data
coap-client -a 127.0.0.1 -p 55555 -m get coap://127.0.0.1/example_data
```

### Code to reproduce this issue

No additional code requried to reproduce

## Debug Logs

Full client and server logs are attached

Interesting excerpts below from server log:

**First GET**
* lg_xmit initialized
* This lg_xmit never gets freed until the session times out 5 minutes after the last message

```
Dec 16 13:35:02.596 DEBG ***[::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : new incoming session
Dec 16 13:35:02.596 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : received 18 bytes
v:1 t:CON c:GET i:052e {01} [ Uri-Path:example_data ]
Dec 16 13:35:02.596 DEBG call custom handler for resource 'example_data'
Dec 16 13:35:02.596 DEBG PDU presented by app.
v:1 t:ACK c:2.05 i:052e {01} [ Content-Format:text/plain ] :: 'a123456789b123456789c ...
Dec 16 13:35:02.596 DEBG ** [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : lg_xmit 0x55fac68a4920 initialized
Dec 16 13:35:02.596 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : sent 1038 bytes
```

** PUT of modified data **
*  Server updates data successfully.

```
v:1 t:CON c:PUT i:6353 {01000000000003} [ Uri-Path:example_data, Block1:0/M/1024, Size1:1500 ] :: 'a987654321b987654321c987654321...
Dec 16 13:35:10.308 DEBG call custom handler for resource 'example_data'
Dec 16 13:35:10.308 DEBG ** [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : lg_srcv 0x55fac68a5030 initialized
Dec 16 13:35:10.308 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : sent 14 bytes
v:1 t:ACK c:2.31 i:6353 {01000000000003} [ Block1:0/M/1024 ]
...
v:1 t:CON c:PUT i:6354 {02000000000003} [ Uri-Path:example_data, Block1:1/_/1024, Size1:1500 ] :: '654321z987654321...
Dec 16 13:35:10.308 DEBG call custom handler for resource 'example_data'
Dec 16 13:35:10.308 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : sent 11 bytes
v:1 t:ACK c:2.04 i:6354 {02000000000003} [ ]
Dec 16 13:35:10.555 DEBG ** [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : lg_srcv 0x55fac68a5030 released
```

** Second GET **
* Server returns incorrect data
```
Dec 16 13:35:14.979 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : received 18 bytes
v:1 t:CON c:GET i:d3cf {01} [ Uri-Path:example_data ]
Dec 16 13:35:14.980 DEBG call custom handler for resource 'example_data'
Dec 16 13:35:14.980 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : sent 1038 bytes
v:1 t:ACK c:2.05 i:d3cf {01} [ ETag:0x01, Content-Format:text/plain, Block2:0/M/1024, Size2:1500 ] :: 'a123456789...'
Dec 16 13:35:14.980 DEBG *  [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : received 26 bytes
```

** Session time out **
* After five minutes, the session times out and the lg_xmit from the first GET is released
* At this point, next GET will return correct data

```
Dec 16 13:40:14.980 DEBG ** [::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : lg_xmit 0x55fac68a4920 released
Dec 16 13:40:14.980 DEBG ***[::]:5683 <-> [::ffff:127.0.0.1]:55555 UDP : session closed
```

## Other items if possible

Does not work in any configuration.  Originally observed with my application running on ARM embedded processor.  Recreated using libcoap example client/server running under WSL2.

**Comment by mrdeep1** [comment]
Thank you for the detailed investigation and reproduction of the issue.  The `lg_xmit `(and hence the cached value) needs to be cleared at time of resource data update. I will work on a fix.

**Comment by dstabosz** [comment]
Thanks for the quick response. 

In my application, I actually saw this with POST, not GET.  A solution that relies on the resource changing might not work in my case.... Or maybe it will if the POST automatically is counted as a resource update. 

I'm new to CoAP and HTTP.  It may be the case that what I am doing is not typical or not legimate.  No need to accomodate my use-case specifically, but let me describe what is happening in my application, so you can consider whether this case should be factored into the fix.

I am using an RPC API where URIs are in the format ```/api/<service>/<method>``` , with the RPC arguments in the CoAP payload.  The specific RPC call I was using is basically reading a memory buffer.  I first saw this with this sequence:

```
# Read 1000 32-bit words.  response requires block-wise transfer
POST /api/<service>/read   payload = { 'offset': 0, 'count': 1000 }

# Read 10 32-bit words.   Server sends back the response from the prior POST, and never calls the handler
POST /api/<service>/read payload = {'offset':0, 'count' : 10}
```
As a workaround, I changed my client code to recreate its client object each time it gets a blocked response.    The main downside I see is that the old session and ```lg_xmit``` will persist on the server for 5 minutes.  

Thanks

**Comment by dstabosz** [comment]
I forgot to attach logs in my initial post.  Attached them here
[client_log.txt](https://github.com/obgm/libcoap/files/7730365/client_log.txt)
[server_log.txt](https://github.com/obgm/libcoap/files/7730366/server_log.txt)

**Comment by mrdeep1** [comment]
Thanks for the additional information.  lg_xmit was set up to effectively cache the response, so that it could be re-used.  I had missed the fact that I was using a different ephmerial port and hence subsequent requests were being treated differently and so had not seen what your were able to reproduce.

However, lg_xmit should not remain in the cache if the rquest is not GET - but only be there for the blocked response - which is the case with your POST.

Similarily, if the resource is changed (which could just be done by the app itself, rather than by another CoAP request) then the caching lg_xmit needs to be removed.

It may be the best solution to remove the lg_xmit whenever a block transfer has completed in the same way that lg_srcv was done in #740

**Comment by mrdeep1** [comment]
A simple general fix would be
````
diff --git a/src/block.c b/src/block.c
index 0563fbf..854a40c 100644
--- a/src/block.c
+++ b/src/block.c
@@ -995,6 +995,9 @@ coap_handle_request_send_block(coap_session_t *session,
   if (coap_get_block(pdu, COAP_OPTION_BLOCK2, &block)) {
     block_opt = COAP_OPTION_BLOCK2;
   }
+  else {
+    return 0;
+  }
   LL_FOREACH(session->lg_xmit, p) {
     size_t chunk;
     coap_opt_iterator_t opt_iter;
````
I agree that the coding could be smarter, but there also needs be support for Q-BLOCK2 (#611).  Please confirm that this works for you.

This only works (the app handler is called) if the request does not contain a BLOCK2.  It is legitimate for a request to include BLOCK2 to indicate the preferred block size for any response, so further work needs to be done here.

Confusion will continue to reign if 2 requests cause 2 blocked responses to occur and the second request is sent before the the first response has completed from the same client IP/port.  The use of the Request-Tag #786 (unique per request) needs to be properly used here.  I will revisit #786 in light of your POST usage to check all is OK there.

The lg_xmit cannot be safely removed at the end of the successful transmission of the body of the payloads as it is possible that the client may request the re-transmission of the last payload (block).

**Comment by dstabosz** [comment]
This is working well for me.  Thanks.

**Comment by mrdeep1** [comment]
Excellent.  I will continue to work through the corner cases I have thought of.

**Comment by mrdeep1** [comment]
Updated version of suggested code fix above is in #801.

Now had clarification on Request-Tag usage in draft-ietf-core-echo-request-tag, so will be updating #786.  See https://mailarchive.ietf.org/arch/msg/core/P0NB_Trjcg-DfLDAr9lj6GgZkwo/
